jms/src/main/scala/akka/stream/alpakka/jms/JmsConnector.scala

```
@@ -90,6 +102,96 @@ private[jms] trait JmsConnector { this: GraphStageLogic =>
+
+    val deliveryModeOption = findHeader(sendHeaders) { case x: JmsDeliveryMode => x.deliveryMode }
+    val priorityOption = findHeader(sendHeaders) { case x: JmsPriority => x.priority }
+    val timeToLiveInMillisOption = findHeader(sendHeaders) { case x: JmsTimeToLive => x.timeInMillis }
+
+    jmsProducer.send(
+      message,
+      deliveryModeOption.getOrElse(jmsProducer.getDeliveryMode),
+      priorityOption.getOrElse(jmsProducer.getPriority),
+      timeToLiveInMillisOption.getOrElse(jmsProducer.getTimeToLive)
+    )
+  }
+
+  private def findHeader[T](headersDuringSend: Set[JmsHeader])(f: PartialFunction[JmsHeader, T]): Option[T] =
+    headersDuringSend.collectFirst(f)
+
+  private def createMessage(element: JmsMessage): Message =
+    element match {
+
+      case textMessage: JmsTextMessage => jmsSession.session.createTextMessage(textMessage.body)
+
+      case byteMessage: JmsByteMessage =>
+        val newMessage = jmsSession.session.createBytesMessage()
+        newMessage.writeBytes(byteMessage.bytes)
+        newMessage
+
+      case mapMessage: JmsMapMessage =>
+        val newMessage = jmsSession.session.createMapMessage()
+        populateMapMessage(newMessage, mapMessage.body)
+        newMessage
+
+      case objectMessage: JmsObjectMessage => jmsSession.session.createObjectMessage(objectMessage.serializable)
+
+    }
+
+  private def populateMessageProperties(message: javax.jms.Message, properties: Map[String, Any]): Unit =
+    properties.foreach {
+      case (key, v) =>
+        v match {
+          case v: String => message.setStringProperty(key, v)
+          case v: Int => message.setIntProperty(key, v)
+          case v: Boolean => message.setBooleanProperty(key, v)
+          case v: Byte => message.setByteProperty(key, v)
+          case v: Short => message.setShortProperty(key, v)
+          case v: Long => message.setLongProperty(key, v)
+          case v: Double => message.setDoubleProperty(key, v)
```

If the properties contain a value which is not of any of these types, would it make sense to explicitly throw an exception instead of the default MatchError?

---

In general, I would say yes.

However, the Jms producer crashes on any exception, and does not distinguish between retriable errors and non-retriable ones - it doesn't perform any retries. So in the end, the behavior would still be very similar if we introduced a custom exception here. It would make much more sense to introduce a custom exception if the Jms producer stage did retries, and would not try to do those retries if the property value type is of an unexpected type.

(please note:  this is part of the code I moved out of the stage into a separate class - I didn't modify it)

---

@WellingR after giving it some thought, those exceptions can be handy in supervision logic to, so here is a proposal - what do you think? https://github.com/akka/alpakka/pull/1148/commits/a2805765cdd09868f878f2d514788d1914d984b1

---

Looks better!
I had some comments about the exception traits though.

---

👍 I've followed your suggestions :)